src/main/java/org/jabref/logic/importer/fileformat/ACMPortalParser.java

```diff
@@ -0,0 +1,254 @@
+import java.util.Locale;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.jabref.logic.importer.FetcherException;
+import org.jabref.logic.importer.ParseException;
+import org.jabref.logic.importer.Parser;
+import org.jabref.logic.net.URLDownload;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.StandardField;
+import org.jabref.model.entry.types.StandardEntryType;
+
+import com.google.common.base.Charsets;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import org.apache.http.client.utils.URIBuilder;
+
+public class ACMPortalParser implements Parser {
+
+    private static final String DOI_URL = "https://dl.acm.org/action/exportCiteProcCitation";
+    private static final Pattern DIO_HTML_PATTERN = Pattern.compile("<input name=\"(.*?)\"");
+    private static final String ITEM_HTML = "<li class=\"search__item issue-item-container\">";
+    private static final int MAX_ITEM_CNT_PER_PAGE = 20;
+    private static HashMap<String, StandardEntryType> ENTRY_TYPE_MAP;
+
+    public ACMPortalParser() {
+        // Gets the lowercase StandardEntryType keys and corresponding StandardEntryType
+        StandardEntryType[] standardEntryTypes = StandardEntryType.values();
+        ENTRY_TYPE_MAP = new HashMap<>(standardEntryTypes.length);
+        for (StandardEntryType standardEntryType : standardEntryTypes) {
+            ENTRY_TYPE_MAP.put(standardEntryType.getName(), standardEntryType);
+        }
+    }
+
+    /**
+     * Parse the DOI of the ACM Portal search result page and obtain the corresponding BibEntry
+     *
+     * @param stream html stream
+     * @return BibEntry List
+     */
+    @Override
+    public List<BibEntry> parseEntries(InputStream stream) throws ParseException {
+        List<BibEntry> bibEntries;
+        try {
+            bibEntries = getBibEntriesFromDoiList(this.parseDoiSearchPage(stream));
+        } catch (FetcherException e) {
+            throw new ParseException(e);
+        }
+        return bibEntries;
+    }
+
+    /**
+     * Parse all DOIs from the ACM Portal search results page
+     *
+     * @param stream html stream
+     * @return DOI list
+     */
+    public List<String> parseDoiSearchPage(InputStream stream) throws ParseException {
+        List<String> doiList = new ArrayList<>();
+        String htmlLine;
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+            int cnt = 0;
+            while ((htmlLine = in.readLine()) != null && cnt < MAX_ITEM_CNT_PER_PAGE) {
+                if (ITEM_HTML.equals(htmlLine)) {
+                    Matcher matcher = DIO_HTML_PATTERN.matcher(in.readLine());
+                    if (matcher.find()) {
+                        doiList.add(matcher.group(1));
+                        ++cnt;
```

Comment:
It usually makes no difference if you use a loop to have a normal post increment count https://stackoverflow.com/a/6175339/3450689

Reply:
I'm not sure what you mean, is it like this?
```java
for(int i = 0 ; i < MAX_ITEM_CNT_PER_PAGE; ){
    while ((htmlLine = in.readLine()) != null){
        if (ITEM_HTML.equals(htmlLine)) {
            Matcher matcher = DIO_HTML_PATTERN.matcher(in.readLine());
            if (matcher.find()) {
                doiList.add(matcher.group(1));
                ++i;
                break;
            }
        }
    }
}
```

But for the original while loop, a better implementation is:
```java
while ((htmlLine = in.readLine()) != null) {
    if (ITEM_HTML.equals(htmlLine)) {
        Matcher matcher = DIO_HTML_PATTERN.matcher(in.readLine());
        if (matcher.find()) {
            doiList.add(matcher.group(1));
            ++cnt;
            if (cnt >= MAX_ITEM_CNT_PER_PAGE) {
                break;
            }
        }
    }
}
```

Reply:
No i just meant i++ isntead of ++i

Reply:
fixed.
The final effect of i++ and ++i is the same, but their efficiency is different.
However, current compilers can automatically compile a single i++ into ++i.

Reply:
`i++` is more reable. We do not do "premature optimization" (is this from https://www.amazon.de/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882/ref=asc_df_0132350882/?tag=googshopde-21&linkCode=df0&hvadid=310746693071&hvpos=&hvnetw=g&hvrand=3074721778032343418&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=9042238&hvtargid=pla-435472505264&psc=1&th=1&psc=1&tag=&ref=&adgrpid=62150065416&hvpone=&hvptwo=&hvadid=310746693071&hvpos=&hvnetw=g&hvrand=3074721778032343418&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=9042238&hvtargid=pla-435472505264?)

Reply:
Understand, fixed.